### ISSUE TITLE
[Bug]: Search failed: node not found[node=0] and collection re-load failed after recovering from oomkilled

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#28772]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20231127-b1e0a27f-amd64
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar 
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus==2.3.2.post1.dev9
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. create collection `fouram_rAXplcI3` with 2 shards -> insert 9m-128dim data into 3 partitions
2. flush -> index -> load
3. concurrent search + delete + flush + scene_insert_partition (concurrent insert 200 entities into a random partition, total 10 partitions)
The datanode and querynode are all oomkilled, datanode memory limit 16Gi
![image](https://github.com/milvus-io/milvus/assets/27288593/1ba1206c-13a7-47da-8a83-5ccc016ddffe)

4. Increase memory limit of datanode and querynode, wait until the pods running, get_replicas and search gets error:
```
[2023-11-27 11:59:49,144 - ERROR - fouram]: (api_response) : <MilvusException: (code=901, message=failed to get shard leader for shard collectionID:445928654284783984 channelName:"delete-stable-3-rootcoord-dml_1_445928654284783984v1" seek_position:<channel_name:"delete-stable-3-rootcoord-dml_1_445928654284783984v1" msgID:"\010\002\020\376\037\030\000 \000" msgGroup:"datanode-5-delete-stable-3-rootcoord-dml_1_445928654284783984v1-true" timestamp:445928859749318657 > flushedSegmentIds:445928654293985577 flushedSegmentIds:445928654291385086 flushedSegmentIds:445928654291985201 flushedSegmentIds:445928654285184356 flushedSegmentIds:445928654285184170 flushedSegmentIds:445928654293385456 flushedSegmentIds:445928654294585702 flushedSegmentIds:445928654295385825 flushedSegmentIds:445928654295985945 flushedSegmentIds:445928654296586084 flushedSegmentIds:445928654292585321 flushedSegmentIds:445928654285184547 dropped_segmentIds:445928654287984490 dropped_segmentIds:445928654287384365 dropped_segmentIds:445928654285984143 dropped_segmentIds:445928654284984030 dropped_segmentIds:445928654288584610 dropped_segmentIds:445928654289984842 dropped_segmentIds:445928654289384732 dropped_segmentIds:445928654290584963 dropped_segmentIds:445928654286584258 : node not found[node=0])> (api_request.py:54)
```

5. release collection and re-load collection again
```
collection not loaded[collection=445928654284783984]
```

### Expected Behavior

1. Is datanode using so much memory as expected ?
2. search should succeed after recoverying from oomkilled. Or at least the re-load should succeed and futher search succeed

### Steps To Reproduce

```markdown
- 4am argo: https://argo-workflows.zilliz.cc/archived-workflows/qa/aab213a5-858b-4e84-a9a3-4bef0641ab52?nodeId=test-delete-stable-partition-10

https://argo-workflows.zilliz.cc/archived-workflows/qa/bf2db54e-fb8e-4d87-a679-e259952f3c9d?nodeId=test-delete-stable-partition-9
```

### Milvus Log

pods:
```
delete-stable-3-milvus-datanode-646cb68c55-kcx2d               1/1     Running            0                 38m     10.104.14.19    4am-node18   <none>           <none>
delete-stable-3-milvus-datanode-646cb68c55-wbmc7               1/1     Running            0                 38m     10.104.1.229    4am-node10   <none>           <none>
delete-stable-3-milvus-indexnode-5464d745f7-xznbr              1/1     Running            0                 89m     10.104.4.86     4am-node11   <none>           <none>
delete-stable-3-milvus-mixcoord-6f74d868fd-6rt7p               1/1     Running            0                 89m     10.104.5.89     4am-node12   <none>           <none>
delete-stable-3-milvus-proxy-66fd99444d-2h5q4                  1/1     Running            0                 89m     10.104.14.3     4am-node18   <none>           <none>
delete-stable-3-milvus-querynode-8c9cf8888-cg7c4               1/1     Running            0                 38m     10.104.21.5     4am-node24   <none>           <none>
delete-stable-3-milvus-querynode-8c9cf8888-dhcvv               1/1     Running            0                 38m     10.104.20.225   4am-node22   <none>           <none>
```

**Comment by ThreadDao** [comment]
1. rerun with image `master-20231206-6a86ac0a-amd64`
2. argo: https://argo-workflows.zilliz.cc/archived-workflows/qa/339a9b1e-fa39-40c6-a355-6c4f8146a334?nodeId=test-delete-stable-partition-12
3. results
  3.1 datanode oomkilled
  
![image](https://github.com/milvus-io/milvus/assets/27288593/975e3629-74b9-42c6-81cd-21f8719c5a85)
  3.2 delete and search StatusCode.DEADLINE_EXCEEDED
4. grafana: https://grafana-4am.zilliz.cc/d/uLf5cJ3Ga/milvus2-0?orgId=1&var-datasource=prometheus&var-cluster=&var-namespace=qa-milvus&var-instance=delta-stable-3&var-collection=All&var-app_name=milvus&from=1701850419799&to=1701864652815
5. pods
```
delta-stable-3-etcd-0                                             1/1     Running                           0                3h42m   10.104.20.187   4am-node22   <none>           <none>
delta-stable-3-etcd-1                                             1/1     Running                           0                3h42m   10.104.16.147   4am-node21   <none>           <none>
delta-stable-3-etcd-2                                             1/1     Running                           0                3h42m   10.104.19.73    4am-node28   <none>           <none>
delta-stable-3-milvus-datanode-6dc5fb7568-7gblz                   1/1     Running                           2 (3h14m ago)    3h38m   10.104.34.152   4am-node37   <none>           <none>
delta-stable-3-milvus-datanode-6dc5fb7568-mvgjp                   1/1     Running                           2 (3h9m ago)     3h38m   10.104.32.64    4am-node39   <none>           <none>
delta-stable-3-milvus-indexnode-5f6748f47b-trcq8                  1/1     Running                           0                3h38m   10.104.33.234   4am-node36   <none>           <none>
delta-stable-3-milvus-mixcoord-b64d8f8b7-mhq6g                    1/1     Running                           0                3h38m   10.104.25.53    4am-node30   <none>           <none>
delta-stable-3-milvus-proxy-6669d87bc7-7vlxw                      1/1     Running                           0                3h38m   10.104.32.63    4am-node39   <none>           <none>
delta-stable-3-milvus-querynode-7d55d9bb8-fdg9b                   1/1     Running                           0                3h38m   10.104.6.237    4am-node13   <none>           <none>
delta-stable-3-milvus-querynode-7d55d9bb8-qrwbn                   1/1     Running                           0                3h38m   10.104.30.253   4am-node38   <none>           <none>
delta-stable-3-minio-0                                            1/1     Running                           0                3h42m   10.104.20.186   4am-node22   <none>           <none>
delta-stable-3-minio-1                                            1/1     Running                           0                3h42m   10.104.16.144   4am-node21   <none>           <none>
delta-stable-3-minio-2                                            1/1     Running                           0                3h42m   10.104.19.72    4am-node28   <none>           <none>
delta-stable-3-minio-3                                            1/1     Running                           0                3h42m   10.104.17.114   4am-node23   <none>           <none>
delta-stable-3-pulsar-bookie-0                                    1/1     Running                           0                3h42m   10.104.20.188   4am-node22   <none>           <none>
delta-stable-3-pulsar-bookie-1                                    1/1     Running                           0                3h42m   10.104.16.148   4am-node21   <none>           <none>
delta-stable-3-pulsar-bookie-2                                    1/1     Running                           0                3h42m   10.104.19.76    4am-node28   <none>           <none>
delta-stable-3-pulsar-bookie-init-wfk5j                           0/1     Completed                         0                3h42m   10.104.25.47    4am-node30   <none>           <none>
delta-stable-3-pulsar-broker-0                                    1/1     Running                           0                3h42m   10.104.26.178   4am-node32   <none>           <none>
delta-stable-3-pulsar-proxy-0                                     1/1     Running                           0                3h42m   10.104.31.151   4am-node34   <none>           <none>
delta-stable-3-pulsar-pulsar-init-mqhk5                           0/1     Completed                         0                3h42m   10.104.25.46    4am-node30   <none>           <none>
delta-stable-3-pulsar-recovery-0                                  1/1     Running                           0                3h42m   10.104.28.170   4am-node33   <none>           <none>
delta-stable-3-pulsar-zookeeper-0                                 1/1     Running                           0                3h42m   10.104.20.189   4am-node22   <none>           <none>
delta-stable-3-pulsar-zookeeper-1                                 1/1     Running                           0                3h41m   10.104.16.152   4am-node21   <none>           <none>
delta-stable-3-pulsar-zookeeper-2                                 1/1     Running                           0                3h41m   10.104.19.80    4am-node28   <none>           <none>
```

**Comment by ThreadDao** [comment]
In addition to the problem of datanode oom in the previous comment. There is another problem that needs to be checked: failed to release-load collection after querynode recovered from oomkilled.
```
[2023-12-07 02:27:29,198 -  INFO - fouram]: [Base] Start load collection fouram_sGbyuuCE,replica_number:1,kwargs:{} (base.py:298)
[2023-12-07 02:39:28,337 - ERROR - fouram]: RPC error: [get_loading_progress], <MilvusException: (code=101, message=collection not loaded[collection=446135546169786735])>, <Time:{'RPC start': '2023-12-07 02:39:25.881699', 'RPC error': '2023-12-07 02:39:28.337006'}> (decorators.py:128)
[2023-12-07 02:39:28,338 - ERROR - fouram]: RPC error: [wait_for_loading_collection], <MilvusException: (code=101, message=collection not loaded[collection=446135546169786735])>, <Time:{'RPC start': '2023-12-07 02:29:30.721185', 'RPC error': '2023-12-07 02:39:28.338392'}> (decorators.py:128)
[2023-12-07 02:39:28,338 - ERROR - fouram]: RPC error: [load_collection], <MilvusException: (code=101, message=collection not loaded[collection=446135546169786735])>, <Time:{'RPC start': '2023-12-07 02:27:29.198702', 'RPC error': '2023-12-07 02:39:28.338576'}> (decorators.py:128)
[2023-12-07 02:39:28,340 - ERROR - fouram]: (api_response) : <MilvusException: (code=101, message=collection not loaded[collection=446135546169786735])> (api_request.py:54)
```
​
1. image:  master-20231206-6a86ac0a-amd64
2. argo workflow:  https://argo-workflows.zilliz.cc/archived-workflows/qa/b8a7ee6e-d44d-4aa6-ac7e-9f45786be8a9?nodeId=test-delete-stable-partition-key-8cdc

https://argo-workflows.zilliz.cc/archived-workflows/qa/f98dd861-4565-49fc-b688-99ff3c95fee5?nodeId=test-delete-stable-partition-key-8

3. pods:
```
delta-par-key2-milvus-datanode-565c688698-h9kfd                   1/1     Running                           0                 13h
delta-par-key2-milvus-datanode-565c688698-w7k7k                   1/1     Running                           0                 13h
delta-par-key2-milvus-indexnode-79959cfc7d-68vc8                  1/1     Running                           0                 13h
delta-par-key2-milvus-mixcoord-75757b77d9-xvxdx                   1/1     Running                           2 (9h ago)        13h
delta-par-key2-milvus-proxy-6d88b5b98c-qstcx                      1/1     Running                           0                 13h
delta-par-key2-milvus-querynode-dfd6d4c8b-kwvwh                   1/1     Running                           3 (10h ago)       13h
delta-par-key2-milvus-querynode-dfd6d4c8b-z9kz9                   1/1     Running                           0                 66m
```

**Comment by ThreadDao** [comment]
- image: master-20231213-fe1eeae2
- argo: https://argo-workflows.zilliz.cc/archived-workflows/qa/8279c077-5d19-47e2-b8f6-268574b54db4?nodeId=test-delete-stable-partition-key-15ab
```
[2023-12-14 08:53:14,009 -  INFO - fouram]: [Base] Start load collection fouram_r2IYATPV,replica_number:1,kwargs:{} (base.py:298)
[2023-12-14 09:03:15,108 - ERROR - fouram]: RPC error: [get_loading_progress], <MilvusException: (code=101, message=collection not loaded[collection=446292310275916201])>, <Time:{'RPC start': '2023-12-14 09:03:15.096062', 'RPC error': '2023-12-14 09:03:15.108415'}> (decorators.py:128)
```
- grafana: https://grafana-4am.zilliz.cc/d/uLf5cJ3Ga/milvus2-0?orgId=1&var-datasource=prometheus&var-cluster=&var-namespace=qa-milvus&var-instance=delta-par-key5&var-collection=All&var-app_name=milvus&from=1702533600000&to=1702545000000
pods:
```
delta-par-key5-milvus-datanode-6f54465fc6-gcrqr                   1/1     Running     2 (147m ago)      21h
delta-par-key5-milvus-datanode-6f54465fc6-qhlt6                   1/1     Running     2 (146m ago)      21h
delta-par-key5-milvus-indexnode-975d8d8fb-ckc9l                   1/1     Running     2 (148m ago)      21h
delta-par-key5-milvus-mixcoord-6cc9f8d965-6mjk9                   1/1     Running     3 (110m ago)      21h
delta-par-key5-milvus-proxy-549d9d5c8f-72sdp                      1/1     Running     1 (145m ago)      21h
delta-par-key5-milvus-querynode-7d7ffd85f6-678j7                  1/1     Running     4 (147m ago)      3h2m
delta-par-key5-milvus-querynode-7d7ffd85f6-cpmvj                  1/1     Running     3 (146m ago)      21h
```

**Comment by congqixia** [comment]
should be fix after #29192
could you please verify
/assign @ThreadDao

**Comment by ThreadDao** [comment]
Fixed, I verified in 2.3 `2.3-af54ce9-20231219`
